client/app/visualizations/EditVisualizationDialog.jsx

```diff
@@ -0,0 +1,211 @@
+    this.setState({
+      name,
+      nameChanged: true,
+    });
+  }
+
+  setVisualizationOptions = (options) => {
+    this.setState({ options: extend({}, options) });
+  };
+
+  dismiss() {
+    const { nameChanged, options, originalOptions } = this.state;
+
+    const optionsChanged = !isEqual(cleanAngularProps(options), originalOptions);
+    if (nameChanged || optionsChanged) {
+      Modal.confirm({
+        title: 'Visualization Editor',
+        content: 'Are you sure you want to close the editor without saving?',
+        okText: 'Yes',
+        cancelText: 'No',
+        onOk: () => {
+          this.props.dialog.dismiss();
+        },
+      });
+    } else {
+      this.props.dialog.dismiss();
+    }
+  }
+
+  save() {
+    const { query } = this.props;
+    const { type, name, options } = this.state;
+
+    const visualization = {
+      ...extend({}, this.props.visualization),
+      ...newVisualization(type),
+    };
+
+    visualization.name = name;
+    visualization.options = options;
+    visualization.query_id = query.id;
+
+    if (visualization.id) {
+      recordEvent('update', 'visualization', visualization.id, { type: visualization.type });
+    } else {
+      recordEvent('create', 'visualization', null, { type: visualization.type });
+    }
+
+    this.setState({ saveInProgress: true });
+
+    Visualization.save(visualization).$promise
+      .then((result) => {
+        toastr.success('Visualization saved');
+
+        const index = findIndex(query.visualizations, v => v.id === result.id);
+        if (index > -1) {
+          query.visualizations[index] = result;
+        } else {
+          // new visualization
+          query.visualizations.push(result);
+        }
+        this.props.dialog.close(result);
+      })
+      .catch(() => {
+        toastr.error('Visualization could not be saved');
+        this.setState({ saveInProgress: false });
+      });
+  }
+
+  render() {
+    const { dialog } = this.props;
+    const { type, name, data, options, canChangeType, saveInProgress } = this.state;
+
+    const { Renderer, Editor, getOptions } = registeredVisualizations[type];
+
+    const previewOptions = getOptions(options, data);
+
+    return (
+      <Modal
+        {...dialog.props}
+        wrapClassName="ant-modal-fullscreen"
+        title="Visualization Editor"
+        okText="Save"
+        okButtonProps={{
+          loading: saveInProgress,
+          disabled: saveInProgress,
+        }}
+        onOk={() => this.save()}
+        onCancel={() => this.dismiss()}
+      >
+        <div className="row">
+          <div className="col-md-5">
+            <div className="m-b-15">
+              <label htmlFor="visualization-type">Visualization Type</label>
+              <Select
+                id="visualization-type"
+                className="w-100"
+                disabled={!canChangeType}
+                value={type}
+                onChange={t => this.setVisualizationType(t)}
```

Comment:
Error thrown in this scenario:
1. New query
2. Insert sql, click execute
3. Click "+new visualization"
4. Change select "Visualization Type" to "Table" (rest of charts ok).
5. Error.

Reply:
I was thinking that it's impossible to open Edit Visualization dialog while query is running/loading results, but that's a case :thinking: Anyway, it's fixed now.